portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -13,20 +13,78 @@
+import com.google.gson.Gson;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
+
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+    private final String com = "Comment";
+    private final String ts = "timestamp";
+    private final String txt = "text";
+
+    /** Store each text input from the server as a Comment class, which has 
+      * properties id, text, and timestamp. 
+      */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        int limit = Integer.parseInt(request.getParameter("limit"));
+        Query query = new Query(com).addSort(ts, SortDirection.DESCENDING);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+
+        List<Comment> comments = new ArrayList<>();
+        int i = 0;
+        for (Entity entity : results.asIterable()) {
+            if (i == limit) {
+                break;
+            } 
+            i++;
+            long id = entity.getKey().getId();
+            String text = (String)entity.getProperty(txt);
+            long timestamp = (long)entity.getProperty(ts);
+
+            Comment comment = new Comment(id, text, timestamp);
+            comments.add(comment);
+        }
+
+        Gson gson = new Gson();
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
-  }
+        response.setContentType("application/json");
+        response.getWriter().println(gson.toJson(comments));
+    }
+
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String text = request.getParameter(txt);
+        long timestamp = System.currentTimeMillis();
+
+        Entity commentEntity = new Entity(com);
+        commentEntity.setProperty(txt, text);
+        commentEntity.setProperty(ts, timestamp);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(commentEntity);
+
+        response.sendRedirect("/index.html");
```

Comment:
Is this necessary considering that we are using the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)?

Reply:
From what I gathered from the tutorials, we need to put the comments into Datastore so the comments aren't deleted every time the server is reloaded

Reply:
I'm not following. I would expect the fetch to not cause the page to reload. Is that not the case?

Reply:
I tried commenting out the redirect to index.html page and when I submit the comment, it redirects to a /data blank page which is why the redirect is necessary